polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -52,6 +54,106 @@ class SYCLRaiseHostConstructsPass
+/// namespace \p targetNamespace.
+static bool isMemberFunction(StringRef demangled, StringRef targetNamespace,
+                             StringRef targetType) {
+  // Check demangled == targetNamespace::...::targetType
+  constexpr StringLiteral namespaceQualifier = "::";
+  return demangled.consume_front(targetNamespace) &&
+         demangled.consume_front(namespaceQualifier) &&
+         demangled.consume_back(targetType) &&
+         demangled.ends_with(namespaceQualifier);
+}
+
+/// Returns whether the function demangled by \p demangler is a member of type
+/// \p targetType in namespace \p targetNamespace.
+static bool isMemberFunction(const llvm::ItaniumPartialDemangler &demangler,
+                             StringRef targetNamespace, StringRef targetType) {
+  FailureOr<DemangleResult> demangled =
+      DemangleResult::get([&](char *buf, std::size_t *size) {
+        return demangler.getFunctionDeclContextName(buf, size);
+      });
+  return succeeded(demangled) &&
+         isMemberFunction(static_cast<StringRef>(*demangled), targetNamespace,
+                          targetType);
+}
+
+/// Returns whether the name of the function demangled by \p demangler matches
+/// \p targetName.
+static bool functionNameMatches(const llvm::ItaniumPartialDemangler &demangler,
+                                StringRef targetName) {
+  FailureOr<DemangleResult> demangled =
+      DemangleResult::get([&](char *buf, std::size_t *size) {
+        return demangler.getFunctionBaseName(buf, size);
+      });
+  return succeeded(demangled) &&
+         static_cast<StringRef>(*demangled) == targetName;
+}
+
+/// Calls `partialDemangle` on \p demangler using the name of the function
+/// referenced by \p ref.
+static LogicalResult partialDemangle(llvm::ItaniumPartialDemangler &demangler,
+                                     SymbolRefAttr ref) {
+  StringRef mangledCalleeName = ref.getLeafReference();
+  return failure(demangler.partialDemangle(mangledCalleeName.data()));
+}
+
+/// Calls `partialDemangle` on \p demangler using the name of the function
+/// called by \p op.
+static LogicalResult partialDemangle(llvm::ItaniumPartialDemangler &demangler,
+                                     CallOpInterface op) {
+  // Get the callable
+  auto callable = dyn_cast<SymbolRefAttr>(op.getCallableForCallee());
+  return callable ? partialDemangle(demangler, callable) : failure();
+}
+
+template <typename T> static auto getUsersOfType(Value value) {
+  constexpr auto filter = [](const OpOperand &operand) -> bool {
+    return isa<T>(operand.getOwner());
+  };
+  constexpr auto map = [](const OpOperand &operand) -> T {
+    return cast<T>(operand.getOwner());
+  };
+  return llvm::map_range(llvm::make_filter_range(value.getUsers(), filter),
+                         map);
+}
+
+/// Returns whether \p type is an `llvm.struct` type with name \p className.
+static bool isClassType(Type type, StringRef className) {
```

Comment:
Should we be using this helper in the buffer/accessor pattern, too (around line 141/243)?

Reply:
Done